File: code/samples/python/src/roles/shopping_agent/subagents/shopper/agent.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""An agent responsible for helping the user shop for products.

The agent gathers the user's purchase intent, sends it to the merchant
to find products, and lets the user select one.
"""

from common.retrying_llm_agent import RetryingLlmAgent
from common.system_utils import DEBUG_MODE_INSTRUCTIONS

from . import tools


shopper = RetryingLlmAgent(
    model="gemini-3.1-flash-lite-preview",
    name="shopper",
    max_retries=5,
    instruction="""
    You are an agent responsible for helping the user shop for products.

    %s

    When asked to complete a task, follow these instructions:
    1. Find out what the user is interested in purchasing. Ask clarifying
      questions one at a time (item description, preferences, budget).
    2. Once you have enough information, call 'find_products' with a
      natural language description of what the user wants. The merchant
      will return product options as cart entries.
    3. Present the options to the user. For each option show:
        - Item name (bold)
        - Price with currency symbol
        - Cart ID
      Recommend the best match and explain why.
    4. Ask the user which item they'd like to purchase.
    5. Once they choose, call 'select_cart' with the cart_id.
    6. If successful, hand off to the root_agent for checkout.
    """ % DEBUG_MODE_INSTRUCTIONS,
    tools=[
        tools.find_products,
        tools.select_cart,
    ],
)

File: samples/typescript/src/roles/merchant-agent/agent.ts
/**
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { LlmAgent } from '@google/adk';
import { DEBUG_MODE_INSTRUCTIONS } from '../../common/constants/index.js';
import { findItemsWorkflow, updateCart, initiatePayment, dpcFinish } from './tools.js';

/**
 * Merchant Agent (ADK)
 *
 * A sales assistant agent for merchants. Handles catalog search, cart updates,
 * and payment processing.
 */
export const merchantAgent = new LlmAgent({
  name: 'merchant_agent',
  model: 'gemini-2.5-flash',
  description: 'A sales assistant agent for a merchant.',
  instruction: `You are a merchant sales assistant agent. Your role is to help customers find products, manage their shopping cart, and complete purchases.

Based on the customer's request, select the appropriate tool to call:

1. **findItemsWorkflow**: Use when the customer wants to search for products or when you receive an IntentMandate. This generates product recommendations.

2. **updateCart**: Use when the customer provides a shipping address and you need to update the cart with shipping costs and taxes.

3. **initiatePayment**: Use when the customer wants to complete their purchase and you have a PaymentMandate.

4. **dpcFinish**: Use when you receive a DPC response to finalize the payment.

Important:
- If you detect a PaymentMandate in the request, immediately call initiatePayment
- Only call one tool per request based on the current context
- After the tool returns a result, respond with a brief summary of the outcome. Do not call the same tool again after it has already returned.

${DEBUG_MODE_INSTRUCTIONS}`,
  tools: [findItemsWorkflow, updateCart, initiatePayment, dpcFinish],
});
